src/liballoc/vec.rs

```diff
@@ -854,6 +854,42 @@ impl<T> Vec<T> {
         ptr
     }
 
+    /// Returns a mutable reference to an element, without doing bounds
+    /// checking.
+    ///
+    /// This is generally not recommended, use with caution!
+    /// Calling this method with an out-of-allocation index is *[undefined behavior]*
+    /// even if the resulting reference is not used.
+    ///
+    /// [undefined behavior]: https://doc.rust-lang.org/reference/behavior-considered-undefined.html
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(vec_get_uninit_unchecked, maybe_uninit_extra)]
+    ///
+    /// // Allocate vector big enough for 4 elements.
+    /// let size = 4;
+    /// let mut x = Vec::with_capacity(4);
+    ///
+    /// // Initialize elements via raw pointer writes, then set length.
+    /// unsafe {
+    ///     for i in 0..size {
+    ///         x.get_uninit_unchecked(i).write(i as i32);
+    ///     }
+    ///     x.set_len(size);
+    /// }
+    /// assert_eq!(&*x, &[0, 1, 2, 3]);
+    /// ```
+    #[unstable(feature = "vec_get_uninit_unchecked", issue = "none")]
+    pub fn get_uninit_unchecked(&mut self, index: usize) -> &mut MaybeUninit<T> {
```

Comment:
```suggestion
    pub unsafe fn get_uninit_unchecked(&mut self, index: usize) -> &mut MaybeUninit<T> {
```

There are two ways in which you can violate soundness with this method:

1. You provide `index` where `index >= self.capacity()`.
2. You do `let elem = vec.get_uninit_unchecked(index); *elem = MaybeUninit::uninit(); drop(vec[index]);`.

Reply:
Also, you can have a memory leak if you forget to use `vec.set_len(...)` afterwards in some circumstances; also worth noting in docs.